include/envoy/server/resource_monitor.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/exception.h"
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Server {
+
+// Struct for reporting usage for a particular resource.
+struct ResourceUsage {
+  // Fraction of (resource usage)/(resource limit).
+  double resource_pressure_;
+};
+
+class ResourceMonitor {
+public:
+  virtual ~ResourceMonitor() {}
+
+  /**
+   * Notifies caller of updated resource usage.
+   */
+  class Callbacks {
+  public:
+    virtual ~Callbacks() {}
+
+    /**
+     * Called when the request for updated resource usage succeeds.
+     * @param usage the updated resource usage
+     */
+    virtual void onSuccess(const ResourceUsage& usage) PURE;
```

Comment:
FWIW, given the size of ResourceUsage I would probably pass by value everywhere, unless you see it getting larger in the future?